src/syntax/parser.rs

```diff
@@ -1066,7 +1066,8 @@ fn for_loop(p: &mut Parser) {
     p.assert(SyntaxKind::For);
     pattern(p);
     if p.at(SyntaxKind::Comma) {
-        p.expected("keyword `in` - did you mean to use a destructuring pattern?");
+        p.expected("keyword `in`");
+        p.err_with_hint("did you mean to use a destructuring pattern?");
```

Comment:
A better solution for the future would be a macro maybe? As expected is a void function, chaining any calls is impossible...

Reply:
`p.expected` could perhaps return a mutable reference to the error?

Reply:
Good Idea! I tried this, and got stuck at the following:

The error/SyntaxNode is moved into `self.nodes`, it can't be returned

```rust
let err = SyntaxNode::error(message, "");
//...
self.nodes.push(err);
//...
&mut err
// err is moved already :/
```

Another downside is, that all expected calls without a hint would have to be discarded:

![image](https://github.com/typst/typst/assets/10061519/187f2402-cacb-4b7d-b717-3d407028da26)

but that's easy, just use `_ =`, so the overhead isn't that big.

If you want I can develop a macro to be used instead of the expected/in supplementation with the expected function - similar to bail!(...)

![image](https://github.com/typst/typst/assets/10061519/04052e71-1c48-44f5-9836-50f9377211ca)

Should I do that?

Reply:
the discarding is a bit of a problem. to return the reference, you need to reacquire a mutable reference into the vector, either with `last_mut().unwrap()` (if it's the last one and no skip happened) or by indexing with `[]`.

Reply:
Good point, thanks! That would probably work - but I think it's not worth it with the discard.

I had another idea when talking with a friend about this:

Instead of having multiple variants and overloads (like expected and expected at in the parser, and .at turning a StrResult into a SourceResult, couldn't we introduce a trait, and then implement Into for Strings, Tuples of String and String and Triplets for more, etc.?

Or would that be too confusing to use?

Reply:
Another idea would be to create `expected_with` to not modify the `expected` method, making the new method return the &mut error, making it look like:

```
p.expected(...) // returns nothing
p.expected_with(...).hint(...); // hint returns nothing
// in the future possibly a new method:
p.expected_with(...).hint_with(...).anything_new();
```

How does that sound? It just means every method is duplicated :)

Reply:
> Instead of having multiple variants and overloads (like expected and expected at in the parser, and .at turning a StrResult into a SourceResult, couldn't we introduce a trait, and then implement Into for Strings, Tuples of String and String and Triplets for more, etc.?

I don't see how this alleviates the need for `At`. Fundamentally some results are just a string and some have a span. To go from one to the other we need a span.

An overloaded `expected` method would work, but for `unexpected` it would already need to be different, so I'm not sure whether it's worth it.

I think the current solution with a method on the parser is the best. I would just rename `p.err_with_hint` to `p.hint`.